File: Recursion/01/NthFibonacci.go
package main
import "fmt"


func nthFibonacci(n int) int {

	if n==1 || n==2{
		return n-1
	}
	return nthFibonacci(n-1) + nthFibonacci(n-2)
}

var dp map[int]int

func memFibonacci(n int) int {


	if n==1 || n==2{
		return n-1
	}else if _,ok := dp[n]; !ok{

		dp[n] = memFibonacci(n-1) + memFibonacci(n-2)

	}

	return dp[n]


}

func top_down_fibonacci(n int )int{

	if n==1 || n ==2{

		return n-1
	}
	a := 0
	b := 1 

	for i:=2;i<n;i++{

		a,b  = b,a+b


	}
	return b

}


func main(){

	dp = make(map[int]int)

	fmt.Println(top_down_fibonacci(8))

}

File: Recursion/01/NthFibonacci.java
import java.util.HashMap;

class NthFibonacci{


	public static int nthFibonacci(int n){

		if(n==1 || n==2)return n-1;

		return nthFibonacci(n-1) + nthFibonacci(n-2);
	}

	private static HashMap<Integer,Integer>dp = new HashMap<>();
	public static int memFibonacci(int n){

		if(n==1 || n==2)return n-1;

		if(!dp.containsKey(n))

			dp.put(n,memFibonacci(n-1) + memFibonacci(n-2));
		return dp.get(n);

		
	}

	public static int topDownFibonacci(int n){


		if(n==1 || n==2)return n-1;

		int a =0,b=1;
		for(int i=2;i<n;i++){

			a = b;
			b = a+b;
		}
		return b;
	}


	public static void main(String[] args) {
		System.out.println(memFibonacci(8));
	}

}